python/foxglove-sdk/python/foxglove/notebook/FoxgloveViewer.py

```diff
@@ -0,0 +1,155 @@
+import importlib.metadata
+import pathlib
+from typing import Any, Optional
+
+import anywidget
+import traitlets
+
+from .._foxglove_py import Context
+from .NotebookBuffer import NotebookBuffer
+
+try:
+    __version__ = importlib.metadata.version("foxglove")
+except importlib.metadata.PackageNotFoundError:
+    __version__ = "unknown"
+
+
+class FoxgloveViewer(anywidget.AnyWidget):
+    """
+    A Jupyter notebook widget that embeds the Foxglove visualization app for interactive
+    data exploration.
+
+    This widget provides a fully-featured Foxglove interface directly within Jupyter notebooks,
+    allowing you to visualize multi-modal robotics data including 3D scenes, plots, images,
+    and more.
+
+    Attributes:
+        width (str): The width of the widget. Defaults to "100%".
+        height (str): The height of the widget. Defaults to "500px".
+        src (str): The URL of the Foxglove app instance to use.
+            If empty, uses the default embed server.
+        layout_data (dict): A custom layout configuration exported from the Foxglove app.
+
+    Example:
+        >>> import foxglove
+        >>> foxglove.create_notebook_buffer()
+        >>> # ... log some data to the buffer ...
+        >>> viewer = foxglove.visualize(width="800px", height="600px")
+        >>> viewer  # Display the widget in the notebook
+    """
+
+    _esm = pathlib.Path(__file__).parent / "static" / "widget.js"
+    _data = traitlets.Bytes(b"").tag(sync=True)
```

Comment:
did we ever explore a way to send this data to the frontend without also holding it permanently in memory on the python side?

Reply:
I tried implementing this, but I ended up finding a bug when sending messages from JS to Python (which I need for synchronization). I filed a [bug](https://github.com/manzt/anywidget/issues/895), and will come back to this when it is fixed.